src/components/TableOfContents.js

```diff
@@ -1,36 +1,41 @@
-import React from "react";
+import React, { useEffect, useState } from 'react';
 
-const link = (title) => `#${title.toLowerCase().replaceAll(' ', '-')}`;
+const slugify = (title) => `${title.toLowerCase().replaceAll(' ', '-')}`;
 
-export class TableOfContents extends React.Component {
-  constructor() {
-    super();
-    this.state = {
-      toc: [],
-    };
-  }
+export function TableOfContents() {
+  const [elements, setElements] = useState([]);
+  const [toc, setToc] = useState([]);
 
-  componentDidMount() {
-    const toc = Array.from(document.querySelectorAll("main > h2")).map(
-      (h2) => h2.innerText
-    );
-    this.setState({ toc });
-  }
+  useEffect(() => {
+    if (elements.length) return;
+    setElements(document.querySelectorAll('main > h2'));
+  });
```

Comment:
If I understand correctly, this will execute 1x and gather up the h2 elements in state for later use...